A setting to change the language of the game to Any language

JackTriton opened the pull request:
A setting to play the randomized oot in Japanese.


Below this is the file informations.

>**textJP.py**
_Raw text of OOT JProm with some transformation for texts_

>**data/titleJP.bin**
_title.bin for JP rom title screen_


Below this is the major definitions I made.

>**parsejp(text, mode = 0)**
_parse the Encoded JPtext (mode 0 = tag, 1 = ending)_

>**text_to_bytes(text, width, signed=False)**
_convert Encoded JPtext to int and to an array of bytes of the given width_

>**JPencode(text, mode = 0, replace_chars=True)**
_encode JPtext written in Japanese and additional CONTROL_CHARS (mode = 0 normal, 1 scrub, 2 item)
when its mode is 2, CONTROL_CHARS are written differently due to icon bytes_

>**update_message_jp(messages, id, text, opts=None, mode = 0)**
_write Encoded JPtexts to temporal.py (mode = 0 normal, 1 scrub, 2 item)
template: (id{:04x}): ((Encoded JPtext), (opts), (tags), (ending)),_

>**shuffle_messages_jp(messages, except_hints=True, always_allow_skip=True)**
_mostly identical function to def shuffle_messages. (return is shuffled groups)_

>**write_messages(rom, shuffle = False, shuffle_group = None)**
_write all texts inside temporal.py to the rom (when shuffle = True, need shuffle_messages_jp)_

I already played several times to ensure that this one doesn't have any sort of unintentional glitch or crashes caused by that.
But if you find it, please comment in this section.

JackTriton committed:
Add files via upload

JackTriton committed:
Add files via upload

JackTriton committed:
Add files via upload

JackTriton committed:
Add files via upload

JackTriton committed:
Update README.md

JackTriton committed:
Update README.md

aofengen commented:
Why is the overall Readme of the project modified to include information from a single PR? Seems unnecessary

cjohnson57 commented:
Just wanna say this is a huge amount of work, great job on it!

JackTriton committed:
Update SettingsList.py

JackTriton committed:
Add files via upload

DuskTheUmbreon commented:
Having a separate Hints.py and Patches.py for every language we'll add in the future doesn't seem terribly sustainable or maintainable (although I could be wrong), especially since we'll likely be adding several other languages after this PR (Almost certainly French, German, Spanish, and Chinese, based on what I've seen in the server). Since both of those files are less "Add this text here" and more "Do this thing in the code", is it possible to just have one file for each?

Obviously we can't condense hintslist.py and hintslistJP.py (or any other file that's 90% "Insert this text into the ROM") into a single file since that'll be a trillion times long by the time we're done with it, but I'd assume we could condense Hints.py and hintsjp.py, and patches.py and patchesJP.py into a single file each.

JackTriton commented:
> Having a separate Hints.py and Patches.py for every language we'll add in the future doesn't seem terribly sustainable or maintainable (although I could be wrong), especially since we'll likely be adding several other languages after this PR (Almost certainly French, German, Spanish, and Chinese, based on what I've seen in the server). Since both of those files are less "Add this text here" and more "Do this thing in the code", is it possible to just have one file for each?
> 
> Obviously we can't condense hintslist.py and hintslistJP.py (or any other file that's 90% "Insert this text into the ROM") into a single file since that'll be a trillion times long by the time we're done with it, but I'd assume we could condense Hints.py and hintsjp.py, and patches.py and patchesJP.py into a single file each.

With using the method I use to create textJP.py, I think we can make same kind of files with all sort of languages.

The reason why I separated to 2 files and didn’t condense (such as patches.py and patchesJP.py) is because in the future, when we’ll add more languages inside this,I’m sure that the file is going to be hard to load or edit it.

DuskTheUmbreon commented:
> > Having a separate Hints.py and Patches.py for every language we'll add in the future doesn't seem terribly sustainable or maintainable (although I could be wrong), especially since we'll likely be adding several other languages after this PR (Almost certainly French, German, Spanish, and Chinese, based on what I've seen in the server). Since both of those files are less "Add this text here" and more "Do this thing in the code", is it possible to just have one file for each?
> > Obviously we can't condense hintslist.py and hintslistJP.py (or any other file that's 90% "Insert this text into the ROM") into a single file since that'll be a trillion times long by the time we're done with it, but I'd assume we could condense Hints.py and hintsjp.py, and patches.py and patchesJP.py into a single file each.
> 
> With using the method I use to create textJP.py, I think we can make same kind of files with all sort of languages.
> 
> The reason why I separated to 2 files and didn’t condense (such as patches.py and patchesJP.py) is because in the future, when we’ll add more languages inside this,I’m sure that the file is going to be too big to load or edit it.

The problem with that is if we make a change to something in patches.py that has nothing to do with the languages at all (say...changing a flag from a temp to a perm), we then need to make that same change another dozen times.

It might be worth splitting patches.py into two files - languagepatches.py and patches.py? That way the majority of the code changes get made to a single file, while the language-specific ones are all separate? I don't know if that's a direction the dev team wants to go in, but it'd solve the issue of having to duplicate non-language changes across every file.

JackTriton commented:
> > > Having a separate Hints.py and Patches.py for every language we'll add in the future doesn't seem terribly sustainable or maintainable (although I could be wrong), especially since we'll likely be adding several other languages after this PR (Almost certainly French, German, Spanish, and Chinese, based on what I've seen in the server). Since both of those files are less "Add this text here" and more "Do this thing in the code", is it possible to just have one file for each?
> > > Obviously we can't condense hintslist.py and hintslistJP.py (or any other file that's 90% "Insert this text into the ROM") into a single file since that'll be a trillion times long by the time we're done with it, but I'd assume we could condense Hints.py and hintsjp.py, and patches.py and patchesJP.py into a single file each.
> > 
> > 
> > With using the method I use to create textJP.py, I think we can make same kind of files with all sort of languages.
> > The reason why I separated to 2 files and didn’t condense (such as patches.py and patchesJP.py) is because in the future, when we’ll add more languages inside this,I’m sure that the file is going to be too big to load or edit it.
> 
> The problem with that is if we make a change to something in patches.py that has nothing to do with the languages at all (say...changing a flag from a temp to a perm), we then need to make that same change another dozen times.
> 
> It might be worth splitting patches.py into two files - languagepatches.py and patches.py? That way the majority of the code changes get made to a single file, while the language-specific ones are all separate? I don't know if that's a direction the dev team wants to go in, but it'd solve the issue of having to duplicate non-language changes across every file.

I think adding language definition to patch_rom itself would be better.
Something like: patch_rom(spoiler, world, rom, lang)
And for the texts to add: 

>if lang is (“English”, “Japanese” etc..): 
    (Excute message update thing)
elif lang is ... 

something like that would solve every problem.

cjohnson57 commented:
You could also make something like a Text class for all text that's inserted into the game, where the class has either an array of text for each language or just an English, Japanese etc. property. HintList instead of taking a string as the hint text will take a Text object that defines both languages, and it decides whether to use the English or Japanese text at patch time depending on the setting.

JackTriton committed:
Add files via upload

JackTriton commented:
I tried my best to compile 2 files to 1 (Like Patches.py and PatchesJP.py).
works fine for most part. 
but when it comes to shop items, it doesn't work at all.
it doesn't place the item (now it's blank)
and soon after that, the game place error text.
I don't know how to fix that. 
so, I'm just going leave that for now (at least) .
But anyway, I fixed some texts in Japanesed V1.2 
and I actually managed to remove about 70% of text speed differences.

To Download: _Japanesed V1.2_ inside the Releases of the branch

JackTriton committed:
Delete LocationTransJP.py

JackTriton committed:
Add files via upload

JackTriton committed:
Add files via upload

JackTriton committed:
Add files via upload

JackTriton committed:
Update version.py

JackTriton committed:
Fetch Up stream

JackTriton committed:
Add files via upload

JackTriton committed:
Add files via upload

JackTriton committed:
Add files via upload

JackTriton committed:
Merge branch 'Dev' into Japanese

JackTriton commented:
Changes:
1. Fill all hint list exclude messages.
2. Fix "Go to message" issue

JackTriton committed:
Add files via upload

JackTriton committed:
Update HintsJP.py

JackTriton committed:
Add files via upload

JackTriton committed:
Add files via upload

Hamsda commented:
> You could also make something like a Text class for all text that's inserted into the game, where the class has either an array of text for each language or just an English, Japanese etc. property. HintList instead of taking a string as the hint text will take a Text object that defines both languages, and it decides whether to use the English or Japanese text at patch time depending on the setting.

I'm pretty sure this is the route we would have to take in the long run. All translatable text needs to be separated from logic bearing things, so this doesn't turn into a maintainability nightmare.

JackTriton committed:
Add files via upload

Succeed in combining Patches.py and PatchesJP.py to one.

JackTriton commented:
> > You could also make something like a Text class for all text that's inserted into the game, where the class has either an array of text for each language or just an English, Japanese etc. property. HintList instead of taking a string as the hint text will take a Text object that defines both languages, and it decides whether to use the English or Japanese text at patch time depending on the setting.
> 
> I'm pretty sure this is the route we would have to take in the long run. All translatable text needs to be separated from logic bearing things, so this doesn't turn into a maintainability nightmare.

So I tried again to compile 2 files to 1, and for the Patches.py, it actually worked!
next time, I'll do the same thing to the Hints.py.

JackTriton committed:
Delete PatchesJP.py

JackTriton committed:
Update Patches.py

JackTriton committed:
Merge branch 'TestRunnerSRL:Dev' into Japanese

JackTriton committed:
Add files via upload

JackTriton committed:
Merge branch 'TestRunnerSRL:Dev' into Japanese

JackTriton committed:
Add support for text random

JackTriton committed:
Add files via upload

JackTriton committed:
Update HintListJP.py

JackTriton committed:
Update Main.py

JackTriton committed:
Merge branch 'Dev' into Japanese

JackTriton committed:
Merge branch 'TestRunnerSRL:Dev' into Japanese

JackTriton committed:
Merge branch 'TestRunnerSRL:Dev' into Japanese

JackTriton committed:
Merge branch 'TestRunnerSRL:Dev' into Japanese

JackTriton committed:
Merge some files

JackTriton committed:
Add files via upload

JackTriton committed:
Delete MessagesJP.py

JackTriton committed:
Delete HintListJP.py

JackTriton committed:
Delete HintsJP.py

JackTriton committed:
Delete TransJP.py

JackTriton commented:
Hi! It's been a while,
I finally managed to compile Messages, Hints and HintList with its JP versions!
In addition, I made line wrap program and a program to align the JP text to center, left and right!
for texts, it is now fully supported for shuffle text and warp song.

JackTriton committed:
Merge branch 'TestRunnerSRL:Dev' into Japanese

JackTriton committed:
Merge branch 'TestRunnerSRL:Dev' into Japanese

JackTriton committed:
Fix Ice Trap

Fix Ice trap shop text for JP and make create fake name

JackTriton committed:
Update Patches.py

JackTriton committed:
Update TextBox.py

JackTriton committed:
Add files via upload

JackTriton committed:
Merge branch 'TestRunnerSRL:Dev' into Japanese

JackTriton committed:
Merge branch 'TestRunnerSRL:Dev' into Japanese

JackTriton committed:
Merge branch 'TestRunnerSRL:Dev' into Japanese

JackTriton committed:
Fix ice trap text

JackTriton committed:
Update Messages.py

JackTriton committed:
Update Messages.py

JackTriton committed:
Merge branch 'TestRunnerSRL:Dev' into Japanese

JackTriton committed:
Merge branch 'TestRunnerSRL:Dev' into Japanese

JackTriton committed:
Fix load glitch

JackTriton committed:
Merge branch 'TestRunnerSRL:Dev' into Japanese

JackTriton committed:
Merge branch 'TestRunnerSRL:Dev' into Japanese

JackTriton committed:
Update texts

Align some JP texts to center

JackTriton committed:
Merge branch 'TestRunnerSRL:Dev' into Japanese

JackTriton committed:
Merge branch 'TestRunnerSRL:Dev' into Japanese

JackTriton committed:
Update texts

Disable prevent skips for texts and rebuild create fake name jp

JackTriton committed:
Fix some texts

Fix some item texts to align it center

JackTriton committed:
Add fake words

Add some fake words for JP ice trap texts

JackTriton committed:
Update Patches.py

JackTriton committed:
Update Patches.py

JackTriton committed:
Update Patches.py

JackTriton committed:
Fix altar texts

Fix weirdly shifted altar texts and add option custom align to update message JP and add few fake words

JackTriton committed:
Update Hints.py

JackTriton committed:
Add files via upload

JackTriton committed:
Change JP logo

Make title logo more look like original randomizer's logo

JackTriton committed:
Update HintList.py

JackTriton committed:
Fix several glitches

JackTriton committed:
Add files via upload

JackTriton committed:
Add files via upload

JackTriton committed:
Add files via upload

JackTriton committed:
Use diff file for titleJP

Use diff file for titleJP and include files that allows you to make title diff file from raw title data file

JackTriton committed:
Add files via upload

JackTriton committed:
Delete titleJP.bin

JackTriton committed:
Rename make title diff.py to make diff.py

JackTriton committed:
Language update

JackTriton committed:
Add files via upload

JackTriton committed:
Merge pull request #2 from JackTriton/Language

Language

JackTriton commented:
I made some codes to get texts from other files using dynamic import.

These are new codes (and files) I added.

data/**make diff.py**
Make diff file from 2 data (original data and changed data)

**ListE.py**
Example table for extra language text file

Below this is the tree structure of extra language file.

(Language File)
|-----ListX.py
|-----data
        |--------title.bin
        |--------etc...
![screen shot](https://user-images.githubusercontent.com/35764777/146287556-a3bb6cd6-145f-4174-82af-8af12b3da1cb.png)
![screen shot (1)](https://user-images.githubusercontent.com/35764777/146287566-b6883f83-792e-4bfc-9a4c-d6b1b441b580.png)

JackTriton committed:
Update SettingsList.py

JackTriton committed:
Update Messages.py

JackTriton committed:
Delete titleoriginal.bin

JackTriton committed:
Merge branch 'Dev' into Japanese

JackTriton committed:
Add files via upload

JackTriton committed:
Merge branch 'Dev' into Japanese

JackTriton committed:
Merge branch 'Dev' into Japanese

fenhl commented:
The approach of passing in a language data file won't work for most languages. There is already some code in this PR that fixes what I assume are grammatical errors arising from simply pasting bits of Japanese text together, and many other languages would require similar considerations. (English is fairly light on those requirements, but even there we already do things like changing “Ganon's castle” to “my castle” in text spoken by Ganondorf.) I think it would be better to remove the “extra language” feature and instead encourage people who want to translate the randomizer to contribute code or collaborate with contributors to make things work. To keep the code organized, I would recommend placing all the code that generates user-visible text in one central module, moving the text itself (including the English text) to JSON files in the `data` directory, and avoiding handling anything language-specific anywhere else.

There is also the concern of what to do when a new feature is added by a contributor who doesn't speak Japanese. Expecting the entire randomizer to be fully translated at all times could significantly hinder development (especially if more languages are added in the future), so instead there should be automatic fallbacks to English text for every missing translation, perhaps combined with a way to query the randomizer for missing translations so that volunteers can get an overview of text to add.

fenhl commented:
It might make sense to combine such a refactor of user-facing text for translation support with the refactor mentioned in https://github.com/TestRunnerSRL/OoT-Randomizer/issues/1525#issuecomment-1060058818.

cjohnson57 commented:
Unfortunately this PR is unlikely to be merged in its current form, but we're going to be making an issue for tracking a translation project, and a lot of the work done here is definitely helpful for that project.

fenhl commented:
Closing in favor of #2441.
